doc/neps/nep-0023-backwards-compatibility.rst

```diff
@@ -19,46 +19,206 @@ processes for individual cases where breaking backwards compatibility
+that NumPy has millions of users, so "no one will do or use this" is very
+likely incorrect.
+
+Benefits include improved functionality, usability and performance, as well as
+lower maintenance cost and improved future extensibility.
+
+Fixes for clear bugs are exempt from this backwards compatibility policy.
+However in case of serious impact on users (e.g. a downstream library doesn't
+build anymore or would start giving incorrect results), even bug fixes may have
+to be delayed for one or more releases.
+
+
+Strategies related to deprecations
+----------------------------------
+
+Getting hard data on the impact of a deprecation of often difficult. Strategies
+that can be used to assess such impact include:
+
+- Use a code search engine ([1]_) or static ([2]_) or dynamic ([3]_) code
+  analysis tools to determine where and how the functionality is used.
+- Testing prominent downstream libraries against a development build of NumPy
+  containing the proposed change to get real-world data on its impact.
+- Making a change in master and reverting it, if needed, before a release. We
+  do encourage other packages to test against NumPy's master branch, so this
+  often turns up issues quickly.
+
+If the impact is unclear or significant, it is often good to consider
+alternatives to deprecations. For example, discouraging use in documentation
+only, or moving the documentation for the functionality to a less prominent
+place or even removing it completely. Commenting on open issues related to it
+that they are low-prio or labeling them as "wontfix" will also be a signal to
+users, and reduce the maintenance effort needing to be spent.
+
+
+Implementing deprecations and removals
+--------------------------------------
+
+Deprecation warnings are necessary in all cases where functionality
+will eventually be removed.  If there is no intent to remove functionality,
+then it should not be deprecated either. A "please don't use this" in the
+documentation or other type of warning should be used instead.
+
+Deprecations:
+
+- shall include the version number of the release in which the functionality
+  was deprecated.
+- shall include information on alternatives to the deprecated functionality, or a
+  reason for the deprecation if no clear alternative is available.
+- shall use ``VisibleDeprecationWarning`` rather than ``DeprecationWarning``
+  for cases of relevance to end users. For cases only relevant to
+  downstream libraries, a regular ``DeprecationWarning`` is fine.
+  *Rationale: regular deprecation warnings are invisible by default; library
+  authors should be aware how deprecations work and test for them, but we can't
+  expect this from all users.*
+- shall be listed in the release notes of the release where the deprecation is
+  first present.
+- shall not be introduced in micro (or bug fix) releases.
+- shall set a ``stacklevel``, so the warning appears to come from the correct
```

Comment:
Should we provide guidance on how this should be done?

Reply:
there's an example further down already